server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,338 @@
+        this.unknownSnapshotShards  = new LinkedHashSet<>();
+        this.queue = new LinkedBlockingQueue<>();
+        this.mutex = new Object();
+        this.activeFetches = 0;
+        this.maxConcurrentFetches = INTERNAL_SNAPSHOT_INFO_MAX_CONCURRENT_FETCHES_SETTING.get(settings);
+        final ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        clusterSettings.addSettingsUpdateConsumer(INTERNAL_SNAPSHOT_INFO_MAX_CONCURRENT_FETCHES_SETTING, this::setMaxConcurrentFetches);
+        if (DiscoveryNode.isMasterNode(settings)) {
+            clusterService.addListener(this);
+        }
+    }
+
+    private void setMaxConcurrentFetches(Integer maxConcurrentFetches) {
+        this.maxConcurrentFetches = maxConcurrentFetches;
+    }
+
+    @Override
+    public SnapshotShardSizeInfo snapshotShardSizes() {
+        return new SnapshotShardSizeInfo(knownSnapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> onGoingSnapshotRecoveries = listOfSnapshotShards(event.state());
+
+            int unknownShards = 0;
+            synchronized (mutex) {
+                isMaster = true;
+                for (SnapshotShard snapshotShard : onGoingSnapshotRecoveries) {
+                    // check if already populated entry
+                    if (knownSnapshotShardSizes.containsKey(snapshotShard) == false) {
+                        // check if already fetching snapshot info in progress
+                        if (unknownSnapshotShards.add(snapshotShard)) {
+                            queue.add(snapshotShard);
+                            unknownShards += 1;
+                        }
+                    }
+                }
+                // Clean up keys from knownSnapshotShardSizes that are no longer needed for recoveries
+                cleanUpKnownSnapshotShardSizes(onGoingSnapshotRecoveries);
+            }
+
+            final int nbFetchers = Math.min(unknownShards, maxConcurrentFetches);
+            for (int i = 0; i < nbFetchers; i++) {
+                fetchNextSnapshotShard();
+            }
+        } else if (event.previousState().nodes().isLocalNodeElectedMaster()) {
+            // TODO Maybe just clear out non-ongoing snapshot recoveries is the node is master eligible, so that we don't
+            // have to repopulate the data over and over in an unstable master situation?
+            synchronized (mutex) {
+                // information only needed on current master
+                knownSnapshotShardSizes = ImmutableOpenMap.of();
+                isMaster = false;
+            }
+        }
+    }
+
+    private void fetchNextSnapshotShard() {
+        synchronized (mutex) {
+            if (activeFetches < maxConcurrentFetches) {
+                try {
+                    final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
```

Comment:
We can save some code here now that we have the `mutex` and don't need a concurrent collection any longer?

e.g.

```diff
diff --git a/server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java b/server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java
index 43cd8cd11b1..a1917c3544b 100644
--- a/server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java
@@ -46,13 +46,12 @@ import org.elasticsearch.repositories.Repository;
 import org.elasticsearch.threadpool.ThreadPool;
 
 import java.util.Collections;
+import java.util.Deque;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
+import java.util.LinkedList;
 import java.util.Objects;
 import java.util.Set;
-import java.util.concurrent.BlockingQueue;
-import java.util.concurrent.LinkedBlockingQueue;
-import java.util.concurrent.TimeUnit;
 import java.util.function.Supplier;
 
 public class InternalSnapshotsInfoService implements ClusterStateListener, SnapshotsInfoService {
@@ -81,7 +80,7 @@ public class InternalSnapshotsInfoService implements ClusterStateListener, Snaps
     private final Set<SnapshotShard> unknownSnapshotShards;
 
     /** a blocking queue used for concurrent fetching **/
-    private final BlockingQueue<SnapshotShard> queue;
+    private final Deque<SnapshotShard> queue;
 
     private volatile int maxConcurrentFetches;
     private volatile int activeFetches;
@@ -99,7 +98,7 @@ public class InternalSnapshotsInfoService implements ClusterStateListener, Snaps
         this.rerouteService = rerouteServiceSupplier;
         this.knownSnapshotShardSizes = ImmutableOpenMap.of();
         this.unknownSnapshotShards  = new LinkedHashSet<>();
-        this.queue = new LinkedBlockingQueue<>();
+        this.queue = new LinkedList<>();
         this.mutex = new Object();
         this.activeFetches = 0;
         this.maxConcurrentFetches = INTERNAL_SNAPSHOT_INFO_MAX_CONCURRENT_FETCHES_SETTING.get(settings);
@@ -159,15 +158,10 @@ public class InternalSnapshotsInfoService implements ClusterStateListener, Snaps
     private void fetchNextSnapshotShard() {
         synchronized (mutex) {
             if (activeFetches < maxConcurrentFetches) {
-                try {
-                    final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
-                    if (snapshotShard != null) {
-                        threadPool.generic().execute(new FetchingSnapshotShardSizeRunnable(snapshotShard));
-                        activeFetches += 1;
-                    }
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    logger.warn("snapshot shard size fetcher has been interrupted", e);
+                final SnapshotShard snapshotShard = queue.pollFirst();
+                if (snapshotShard != null) {
+                    threadPool.generic().execute(new FetchingSnapshotShardSizeRunnable(snapshotShard));
+                    activeFetches += 1;
                 }
             }
             assert assertNumberOfConcurrentFetches();
```

Reply:
Can we just use `poll()` and make the queue declaration non blocking? That also the interrupt handling.

Reply:
Makes perfect sense, thanks for noticing this! I pushed fab7abb83888fd83b75d226e389534b049697dfa